aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -3384,5 +3387,69 @@ Tensor kron(const Tensor& self, const Tensor& other) {
   return KronImpl(self, other).kron();
 }
 
+// Weight Only Quantization Gemm
+DEFINE_DISPATCH(weight_to_int4pack_stub);
+DEFINE_DISPATCH(int4pack_mm_stub);
+
+Tensor _convert_weight_to_int4pack_cpu(
+    const Tensor& in,
+    int64_t innerKTiles) {
+
+  TORCH_CHECK(in.dim() == 2,
+      "_convert_weight_to_int4pack: expect weight to be 2D tensor.");
+  TORCH_CHECK(in.dtype() == at::kInt,
+      "_convert_weight_to_int4pack: expect weight to be kInt.");
+
+  auto weight = in.contiguous();
+  auto N = weight.size(0);
+  auto K = weight.size(1);
+
+  TORCH_CHECK(N % 16 == 0,
+      "_convert_weight_to_int4pack: expect N to be dividable by 16");
+  TORCH_CHECK(K % 2 == 0,
+      "_convert_weight_to_int4pack: expect K to be dividable by 2");
+
+  auto weight_packed = at::empty({N, K / 2}, weight.options().dtype(kByte));
+  weight_to_int4pack_stub(kCPU, weight_packed, weight);
+  return weight_packed;
+}
+
+Tensor _weight_int4pack_mm_cpu(
+    const Tensor& A,
+    const Tensor& B,
+    int64_t qGroupSize,
+    const Tensor& qScaleAndZeros) {
+
+  auto M = A.size(0);
+  auto N = B.size(0);
+  auto K = A.size(1);
+
+  TORCH_CHECK(A.dtype() == kBFloat16,
+      "_weight_int4pack_mm: expect A to be bfloat16 tensor.");
+  TORCH_CHECK(A.is_contiguous(),
+      "_weight_int4pack_mm: expect A to be contiguous.");
+  TORCH_CHECK(A.dim() == 2,
+      "_weight_int4pack_mm: expect A to be 2D tensor.");
+
+  TORCH_CHECK(B.dtype() == kByte,
+      "_weight_int4pack_mm: expect B to be uint8 tensor.");
+  TORCH_CHECK(B.is_contiguous(),
+      "_weight_int4pack_mm: expect B to be contiguous.");
+  TORCH_CHECK(B.size(1) == K / 2);
+
+  TORCH_CHECK(
+      qGroupSize == 32 || qGroupSize == 64 || qGroupSize == 128 ||
+      qGroupSize == 256);
```

Comment:
Nit (also, consider adding some user readable error message (and what is wrong with say 512 groupsize)
```suggestion
      qGroupSize >= 32 && qGroupSize <= 256 || (qGroupSize & (qGroupSize - 1)) == 0);
```

Reply:
the frontend of gpt-fast has constrait for qGroupSize to be 32, 64, 128, 256 now : https://github.com/pytorch-labs/gpt-fast/blob/main/quantize.py#L402. 

larger group size has smaller model size (but the benefit is not big since `ScalesAndZeros` is already much smaller compared to the int4 packed weight) but will have worse accuracy. (group size has little impact on perf)

by default gpt-fast uses group size of 32 which is the same as ggml.